injectable_generator/lib/code_builder/library_builder.dart

```diff
@@ -383,18 +401,21 @@ Expression _getDisposeFunctionAssignment(DisposeFunctionConfig disposeFunction,
 
 Expression _buildParamAssignment(
   InjectedDependency iDep,
+  Set<DependencyConfig> allDeps,
   Uri? targetFile, {
   required String name,
 }) {
   if (iDep.isFactoryParam) {
     return refer(iDep.paramName);
   }
-  return refer(name).call([], {
+  final isAsync = isAsyncOrHasAsyncDependency(iDep, allDeps);
+  final expression = refer(isAsync ? '$name.getAsync' : name).call([], {
```

Comment:
I noticed this won't work when the init function is implemented as an extension method on GetIt. In that case, `$name` refers to the actual `get` method on the GetIt instance, whereas when it is not implemented as an extension method `$name` refers to the GetIt instance itself. So this will need to change a little bit.